src/pybind/mgr/cephadm/serve.py

```diff
@@ -510,50 +530,54 @@ def matches_network(host):
         remove_daemon_hosts: Set[orchestrator.DaemonDescription] = ha.remove_daemon_hosts(hosts)
         self.log.debug('Hosts that will loose daemons: %s' % remove_daemon_hosts)
 
-        for host, network, name in add_daemon_hosts:
-            daemon_id = self.mgr.get_unique_name(daemon_type, host, daemons,
-                                                 prefix=spec.service_id,
-                                                 forcename=name)
-
-            if not did_config and config_func:
-                if daemon_type == 'rgw':
-                    rgw_config_func = cast(Callable[[RGWSpec, str], None], config_func)
-                    rgw_config_func(cast(RGWSpec, spec), daemon_id)
-                else:
-                    config_func(spec)
-                did_config = True
-
-            daemon_spec = self.mgr.cephadm_services[daemon_type].make_daemon_spec(
-                host, daemon_id, network, spec)
-            self.log.debug('Placing %s.%s on host %s' % (
-                daemon_type, daemon_id, host))
+        if service_type == 'ha-rgw':
+            spec = self.update_ha_rgw_definitive_hosts(spec, hosts, add_daemon_hosts)
 
-            try:
-                daemon_spec = self.mgr.cephadm_services[daemon_type].prepare_create(daemon_spec)
-                self.mgr._create_daemon(daemon_spec)
-                r = True
-            except (RuntimeError, OrchestratorError) as e:
-                self.mgr.events.for_service(spec, 'ERROR',
-                                            f"Failed while placing {daemon_type}.{daemon_id}"
-                                            f"on {host}: {e}")
-                # only return "no change" if no one else has already succeeded.
-                # later successes will also change to True
-                if r is None:
-                    r = False
-                continue
+        for host, network, name in add_daemon_hosts:
+            for daemon_type in service_to_daemon_types(service_type):
+                daemon_id = self.mgr.get_unique_name(daemon_type, host, daemons,
+                                                     prefix=spec.service_id,
+                                                     forcename=name)
```

Comment:
interesting, `service_id` starts with `rgw-ha`, what does `daemon_id` look like then here?

Reply:
The log statement a few lines later that prints daemon_type.daemon_id came out like
`Placing haproxy.haproxy_for_rgw.vm-00.yaljjn on host vm-00` for haproxy and
`Placing keepalived.haproxy_for_rgw.vm-00.ymborn on host vm-00` for keepalived 
on host vm-00 coming from a spec file with
`service_type: ha-rgw`
`service_id: haproxy_for_rgw`
It doesn't seem like `ha-rgw` appears in the id anywhere unless you explicitly include it as part of your service id in the spec